Start adding support for Kokkos and the Nvidia compiler

nilsdeppe opened the pull request:
## Proposed changes

My plan is to first get support for nvcc (the Nvidia compiler) to compile our code. nvcc uses the Edison Design Group (EDG) compiler front end, and so will parse code differently from GCC and Clang. I plan on first getting the code to compile with nvcc without target GPUs. Once that's working I'll continue getting the necessary bits to run on a GPU. The functions with `KOKKOS_FUNCTION` in front are ones that are compiled with nvcc to run on the GPU.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add Kokkos wrapper utilities library

nilsdeppe committed:
Add Kokkos support to CMake

nilsdeppe committed:
Move Blaze exception macros into a file

This is needed for interop with Kokkos and nvcc

nilsdeppe committed:
Add nvcc support Utilities and DataStructures

nilsdeppe committed:
Add nvcc support to brigand

nilsdeppe committed:
Add Kokkos support to Error.hpp

nilsdeppe committed:
Add Kokkos support to gsl::at

nilsdeppe committed:
Add Kokkos support to cpp20::array

nilsdeppe committed:
Add Kokkos support to ConstantExpressions

nilsvu commented on tools/BlazeExceptions.hpp:
```diff
@@ -0,0 +1,19 @@
+#define SPECTRE_BLAZE_EXCEPTIONS_HPP
+
+#ifdef SPECTRE_DEBUG
+#define BLAZE_THROW(EXCEPTION)           \
```
Can you just include `csignal` here so it doesn't need special handling in cmake?

nilsvu commented on src/Utilities/Simd/Simd.hpp:
```diff
@@ -242,17 +242,25 @@ T select(const bool cond, const T true_branch, const T false_branch) {
 }
 
 inline std::pair<float, float> sincos(const float val) {
+#if (defined(__CUDACC__) && defined(__CUDA_ARCH__)) or (not defined(__CUDACC__))
```
Can you add a comment when this needs to be disabled? What does it mean if `__CUDA_ARCH__` is undefined?

nilsvu commented on src/DataStructures/Tensor/EagerMath/CartesianToSpherical.cpp:
```diff
@@ -5,6 +5,7 @@
 
 #include <cstddef>
 
+#include "DataStructures/DataVector.hpp"
```
Question: can we limit `nvcc` support to only part of the code (essentially PointwiseFunctions & NumericalAlgorithms) so we don't have to keep the full code compatible with nvcc? I.e., can we compile only GPU kernels with `nvcc` and the rest with clang?

nilsdeppe replied:
Good question! I think practically speaking the answer is "no." Here are a list of things I can immediately think of that I think would be challenging to overcome (in no particular order):
1. We would need to figure out how to have CMake use a different compiler for different libraries.
2. We would need to filter out the compiler arguments that Kokkos adds for nvcc that GCC/Clang can't handle. I'm already running into this with our linker wrapper, but I haven't dug into what's going on there.
3. There may be ABI incompatibilities now or in the future because things like `__host__` and `__device__` may or may not be included in the function signature. This is implementation defined and Clang and nvcc currently behave differently.
4. We need essentially all header files to be compatible with nvcc anyway since they get included in PointwiseFunctions, NumericalAlgorithms, ParallelAlgorithms, and the final executable (I think we need to compile the final executable with nvcc, I don't see a way around that). This means we would only benefit from not having to make the cpp files compatible, which is definitely not negligible, but I worry that this will be very confusing in the long term. I've found that it is also easier to make one library at a time work with nvcc rather than trying to deal with all the possible headers files when building at the end. Essentially, incremental conversion, even of full libraries, seems easier than many libraries all at once from my experience so far.

nilsvu replied:
Looks like CMake supports separate host and device compilers with [CUDA_SEPARABLE_COMPILATION](https://cmake.org/cmake/help/latest/prop_tgt/CUDA_SEPARABLE_COMPILATION.html). I haven't seen support for this in Kokkos though.

We can also try `-DCMAKE_CUDA_COMPILER=clang++` to avoid having to keep compatibility with `nvcc` for the whole code.

nilsdeppe replied:
That might be worth trying. I just don't know how it would work with Kokkos because kokkos actually has a wrapper around the Nvidia compiler.

nilsvu commented on src/Utilities/ErrorHandling/Error.hpp:
```diff
@@ -85,6 +91,11 @@ template <typename ExceptionTypeToThrow, typename F>
  * instantiation in `Utilities/ErrorHandling/AbortWithErrorMessage.cpp`
  */
+#ifdef __CUDA_ARCH__
+#define ERROR_AS(m)                                           \
```
Needs additional `EXCEPTION_TYPE` arg, or will it just be discarded?

nilsdeppe committed:
Add csignal to BlazeExceptions.hpp

This moves we no longer need to explicitly include it.

nilsdeppe commented:
I've rebased and made the changes. I squashed most since they were 1 line, and added a commit for `csignal` since it's logically different from the previous commit. Thanks for the review!

nilsvu commented on docs/Installation/BuildSystem.md:
````diff
@@ -527,3 +529,22 @@ To further aid in reproducibility, the `printenv` output and
 h5dump -d /src.tar.gz -b LE -o src.tar.gz /path/to/hdf5/file.h5
 ```
+
+# Using Kokkos
````
Should we install Kokkos in the container and/or in a CI build? I haven't checked how well we can test (emulated) GPUs on CI, but we can certainly test compiling with Kokkos on CPU. Also compatibility with nvcc would be important to test on CI so we don't accidentally break it.

nilsdeppe replied:
Short answer: yes. What I've been doing is compiling on mbot for the A100 GPUs. I don't need the GPU to compile the code, but presumably I would to run it. I think in CI what we can do is install CUDA, Kokkos, and then at least we can compile with nvcc to make sure we don't break anything. We'll want to do the same for the ROCm and DPC++ compilers too, but that's a more future problem :)
